refactor unit tests to be able to call by deployment type

leelavg opened the pull request:
Signed-off-by: Leela Venkaiah G <lgangava@redhat.com>

leelavg commented:
- My aim for this PR is to separate out make file targets for deploymentType
- Minimal changes to the go code is made not to have huge diff
- Pls let me know if we want to re-use a single envtest for all three mode of deployments (I prefer not)

```
$ make test
[...]
MODE=converged go test ./... -coverprofile cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 100.064s        coverage: 76.6% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    17.505s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
[...]
MODE=provider go test ./... -coverprofile cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 99.760s coverage: 75.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    16.102s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
[...]
MODE=consumer go test ./... -coverprofile cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 94.935s coverage: 72.9% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    16.681s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
```

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -1959,34 +1912,8 @@ var _ = Describe("ManagedOCS controller", func() {
-	runTests(convergedDeploymentType)
-	runTests(consumerDeploymentType)
-	runTests(providerDeploymentType)
+	runTest()
```
Why do we need a function call? Can we just unwrap and put the context here? (the way it was before we added the `runTests` functions?

leelavg replied:
- only a single reason, huge diff due to de-tab after removing function and important changes will not be evident
- I'm thinking to raise another pr just removing this anon function, I can add another commit if that suffices

nb-ohad replied:
Why do you care about the diff size? I would argue that we should care about having each PR focused on a single coherent change - in this case, you are already touching the function and removing all the calls to it, so making the last step of inlining it just make sense

nb-ohad commented on controllers/suite_test.go:
```diff
@@ -77,6 +79,17 @@ const (
+	// check for which deployment type we need to run the tests for
+	envVar := "MODE"
+	testDeployment = os.Getenv(envVar)
+	switch testDeployment {
+	case convergedDeploymentType, providerDeploymentType, consumerDeploymentType:
+	default:
+		panic(fmt.Sprintf("Environment var %s should be set to one of %s, %s and %s values",
+			envVar, convergedDeploymentType, providerDeploymentType, consumerDeploymentType))
+	}
```
You have one condition and one test here so why use `switch` instead of an `if`?
```suggestion
	if testDeployment != convergedDeploymentType &&
		testDeployment != providerDeploymentType &&
		testDeployment != consumerDeploymentType {
		panic(fmt.Sprintf("Environment var %s should be set to one of %s, %s and %s values",
			envVar, convergedDeploymentType, providerDeploymentType, consumerDeploymentType))
	}
```

leelavg replied:
- I usually prefer `switch` for more than two args with same result 
- if you disagree, I'll change it to `if` condition

leelavg replied:
- i can validate all required values in a single `case` statement and `default` will be panic

nb-ohad replied:
I just saw you how to validate using a single if statement. I will argue that because `switch` in go has the opposite semantics from any other C-based language (break is explicit and not fallthrough which is the opposite of any other C-based lang that I am aware of ) then the above is very very misleading for the casual reader as we are so used to have it the other way around

nb-ohad commented on Makefile:
```diff
@@ -42,10 +42,24 @@ export DEPLOYMENT_TYPE = converged
-	source $(ENVTEST_ASSETS_DIR)/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test ./... -coverprofile cover.out
+	source $(ENVTEST_ASSETS_DIR)/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR);
+
+test-converged:
```
```suggestion
test-converged: generate fmt vet manifests
```

nb-ohad commented on Makefile:
```diff
@@ -42,10 +42,24 @@ export DEPLOYMENT_TYPE = converged
+	$(MAKE) -B setup-envtest
+	MODE=converged go test ./... -coverprofile cover.out
+
+test-provider:
```
```suggestion
test-provider: generate fmt vet manifests
```

nb-ohad commented on Makefile:
```diff
@@ -42,10 +42,24 @@ export DEPLOYMENT_TYPE = converged
+	$(MAKE) -B setup-envtest
+	MODE=provider go test ./... -coverprofile cover.out
+
+test-consumer:
```
```suggestion
test-consumer: generate fmt vet manifests
```

nb-ohad commented on Makefile:
```diff
@@ -42,10 +42,24 @@ export DEPLOYMENT_TYPE = converged
+	$(MAKE) -B setup-envtest
+	MODE=consumer go test ./... -coverprofile cover.out
+
+test: generate fmt vet manifests test-converged test-provider test-consumer
```
```suggestion
test: test-converged test-provider test-consumer
```

nb-ohad commented on Makefile:
```diff
@@ -42,10 +42,24 @@ export DEPLOYMENT_TYPE = converged
+	source $(ENVTEST_ASSETS_DIR)/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR);
+
+test-converged:
+	$(MAKE) -B setup-envtest
```
Why not just depend on setup-envtest target?

leelavg replied:
- as there'll no file changes during `setup-envtest` we can end up using same environment 
- since I doubt some resources may be left behind even after `AfterSuite` I just made `setup-envtest` to always run

nb-ohad replied:
But here you are running a single test, so there will be no env test reusing

leelavg commented on controllers/suite_test.go:
```diff
@@ -77,6 +79,17 @@ const (
 func TestAPIs(t *testing.T) {
+
+	// check for which deployment type we need to run the tests for
+	envVar := "MODE"
```
I initially thought of that, even though it's not directly linked can I remove this line https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/Makefile#L41 then?

nb-ohad replied:
I think that line is relevant for `make run`. so we run `DEPLOYMENT_TYPE=provider make run` in order to set a deployment type. But we can use the same var for different make targets

leelavg commented:
- addressed comments and re-ran tests 
```
# make test
/root/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
api/v1alpha1/zz_generated.deepcopy.go
go vet ./...
/root/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
mkdir -p /root/toolbox/orc/ocs-osd-deployer/testbin
test -f /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh || curl -sSLo /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.6.3/hack/setup-envtest.sh
source /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh; fetch_envtest_tools /root/toolbox/orc/ocs-osd-deployer/testbin; setup_envtest_env /root/toolbox/orc/ocs-osd-deployer/testbin;
Using cached envtest tools from /root/toolbox/orc/ocs-osd-deployer/testbin
setting up env vars
DEPLOYMENT_TYPE=converged go test ./... -coverprofile converged-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 100.620s        coverage: 76.0% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    16.929s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
DEPLOYMENT_TYPE=provider go test ./... -coverprofile provider-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 97.558s coverage: 75.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    17.291s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
DEPLOYMENT_TYPE=consumer go test ./... -coverprofile consumer-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 100.140s        coverage: 73.1% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    16.243s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
```

nb-ohad commented on controllers/suite_test.go:
```diff
@@ -77,6 +79,17 @@ const (
 func TestAPIs(t *testing.T) {
+
+	// check for which deployment type we need to run the tests for
+	envVar := "DEPLOYMENT_TYPE"
```
Why not a const in the consts section above? 
Also, this is the name of the env var so can we be more descriptive and call this `deploymentTypeEnvVarName`?

nb-ohad commented on api/v1alpha1/zz_generated.deepcopy.go:
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
```
What is this change and why do we need it?

leelavg replied:
- it seems recent versions of go tools are prefixing `go:` for code generation 
- this shouldn't cause any issue for us

nb-ohad replied:
ok

leelavg commented:
- addressed comments

leelavg commented:
rebased ```
# make test
/root/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
api/v1alpha1/zz_generated.deepcopy.go
go vet ./...
/root/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
mkdir -p /root/toolbox/orc/ocs-osd-deployer/testbin
test -f /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh || curl -sSLo /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.6.3/hack/setup-envtest.sh
source /root/toolbox/orc/ocs-osd-deployer/testbin/setup-envtest.sh; fetch_envtest_tools /root/toolbox/orc/ocs-osd-deployer/testbin; setup_envtest_env /root/toolbox/orc/ocs-osd-deployer/testbin;
Using cached envtest tools from /root/toolbox/orc/ocs-osd-deployer/testbin
setting up env vars
DEPLOYMENT_TYPE=converged go test ./... -coverprofile converged-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 97.429s coverage: 75.7% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    15.822s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
DEPLOYMENT_TYPE=provider go test ./... -coverprofile provider-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 95.818s coverage: 75.0% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    16.759s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
DEPLOYMENT_TYPE=consumer go test ./... -coverprofile consumer-cover.out
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 84.015s coverage: 73.0% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    15.010s coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
```
